projects/core/src/main/resources/data/computercraft/lua/rom/programs/shell.lua

```diff
@@ -54,6 +54,67 @@ else
     bgColour = colours.black
 end
 
+local function tokenise(...)
+    local sLine = table.concat({ ... }, " ")
+    local tWords = {}
+    local bQuoted = false
+    for match in string.gmatch(sLine .. "\"", "(.-)\"") do
+        if bQuoted then
+            table.insert(tWords, match)
+        else
+            for m in string.gmatch(match, "[^ \t]+") do
+                table.insert(tWords, m)
+            end
+        end
+        bQuoted = not bQuoted
+    end
+    return tWords
+end
+
+-- Execute a program using os.run, unless a shebang is present.
+-- In that case, execute the program using the interpreter specified in the hashbang.
+-- This may occur recursively, up to the maximum number of times specified by remainingRecursion
+-- Returns the same type as os.run, which is a boolean indicating whether the program exited successfully.
+local function executeProgram(remainingRecursion, env, path, ...)
+    local file, err = fs.open(path, "r")
+    if not file then
+        printError(err)
+        return false
+    end
+
+    -- First check if the file begins with a #!
+    local contents = file.readLine()
+    file.close()
+
+    if contents:sub(1, 2) == "#!" then
+        remainingRecursion = remainingRecursion - 1
+        if remainingRecursion == 0 then
+            printError("Hashbang recursion depth limit reached when loading file: " .. path)
+            return false
+        end
+
+        -- Load the specified hashbang program instead
+        local args = tokenise(contents:sub(3))
+        local originalHashbangPath = table.remove(args, 1)
+        local resolvedHashbangProgram = shell.resolveProgram(originalHashbangPath)
+        if not resolvedHashbangProgram then
+            printError("Hashbang program not found: " .. originalHashbangPath)
+            return false
+        end
+
+        -- Add the path and any arguments to the interpreter's arguments
+        table.insert(args, path)
+        for _, v in ipairs({ ... }) do
+            table.insert(args, v)
+        end
+
+        env.arg = { [0] = resolvedHashbangProgram, unpack(args) }
+        return executeProgram(remainingRecursion, env, resolvedHashbangProgram, unpack(args))
```

Comment:
```suggestion
        env.arg = { [0] = resolvedHashbangProgram, table.unpack(args) }
        return executeProgram(remainingRecursion, env, resolvedHashbangProgram, table.unpack(args))
```

`unpack` is notionally deprecated, as it's a Lua 5.1-only feature.

One alternative here would be to do what PUC Lua does, and instead of passing varargs through the functions, do `os.run(env, path, table.unpack(env.arg))`. Not sure if it's any nicer though - will leave that up to you!